### PR TITLE
Add rules for concept exercise `treasure-factory`

### LINKED ISSUE #57: Write analysis for concept exercise treasure-factory
From design.md
- [essential] the type signatures of makeChest and makeTreasureChest match the ones in the stub (show them in the comment). This will fix https://github.com/exercism/elm/issues/585.
- [essential] the type signature of secureChest is exactly Chest treasure constraints -> Maybe (Chest treasure { constraints | securePassword : () }) (don't show it in the comment)
- [essential] the type signature of uniqueTreasures is exactly List (Chest treasure constraints) -> List (Chest treasure { constraints | uniqueTreasure : () }) (don't show it in the comment)

### PR DESCRIPTION
**Pull request by jiegillet**
Closes #57

[Sister PR here](https://github.com/exercism/website-copy/pull/2285).
The way to specify signatures is super verbose, but I think it's still the best way to do it.

### CODE REVIEW

File: tests/ElmSyntaxHelpersTest.elm [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+                ElmSyntaxHelpers.typeAnnotationsMatch noRangeUnit someRangeUnit
+                    |> Expect.equal True
+        , fuzz typeAnnotationsFuzzer "a TypeAnnotation matches itself" <|
+            \a ->
+                ElmSyntaxHelpers.typeAnnotationsMatch a a
+                    |> Expect.equal True
+        , fuzz2 typeAnnotationsFuzzer typeAnnotationsFuzzer "only matches equal TypeAnnotations" <|
+            \a b ->
+                ElmSyntaxHelpers.typeAnnotationsMatch a b
+                    |> Expect.equal (a == b)
```
Comment (jiegillet) [inline tests/ElmSyntaxHelpersTest.elm:32]: This silly looking fuzz test showed me a bug straight away 💛 
(I was comparing lists with `List.map2` and forgot to check that they needed to be of the same size)

**Review by ErikSchierboom — APPROVED** [review]

**Review by ceddlyburge — APPROVED** [review]
Nice, this is going to fix 2 issues :)

File: src/Exercise/TreasureFactory.elm [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+
+makeChestSignatureMatchesStub : Comment -> Rule
+makeChestSignatureMatchesStub comment =
+    let
+        expectedSignature =
+            { name = Node.empty "makeChest"
+            , typeAnnotation =
+                Node.empty
```
[the remark this answers is not in the document]
Comment (jiegillet) [reply src/Exercise/TreasureFactory.elm:37]: Agreed :)

File: src/Exercise/TreasureFactory.elm [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+
+signatureMustBe : Comment -> Signature -> Rule
+signatureMustBe comment expectedSignature =
+    Rule.newModuleRuleSchema comment.path []
+        |> Rule.withSimpleDeclarationVisitor (hasExpectedSignatureVisitor comment expectedSignature)
+        |> Rule.fromModuleRuleSchema
+
+
+hasExpectedSignatureVisitor : Comment -> Signature -> Node Declaration -> List (Error {})
```
[the remark this answers is not in the document]
Comment (jiegillet) [reply src/Exercise/TreasureFactory.elm:193]: That's a good idea. As of now, that's the only time it's used, but I'll extract it if it's useful somewhere else

**Review by ErikSchierboom — APPROVED** [review]

**Pull request merged**